File: src/24-challenge/solution.js
/**
 * @typedef {{ value: string; left?: BinaryTree; right?: BinaryTree }} BinaryTree
 */

/**
 * @param {BinaryTree | undefined} tree1 - The first binary tree.
 * @param {BinaryTree | undefined} tree2 - The second binary tree.
 * @returns {[boolean, string]}
 */
export default function isTreesSynchronized(tree1, tree2) {
	/** @type {[boolean, string]} */
	const result = [tree1?.value === tree2?.value, `${tree1?.value}`]
	if (!tree1?.left || !tree2?.right) return result

	/** @type {BinaryTree} */
	const leftBranch01 = tree1.left

	/** @type {BinaryTree | undefined} */
	const rightBranch01 = tree1.right

	/** @type {BinaryTree | undefined} */
	const leftBranch02 = tree2.left

	/** @type {BinaryTree} */
	const rightBranch02 = tree2.right

	result[0] = result[0] && isTreesSynchronized(leftBranch01, rightBranch02)[0]
	result[0] = result[0] && isTreesSynchronized(rightBranch01, leftBranch02)[0]

	return result
}

File: src/24-challenge/solution.py
from typing import TypedDict, NotRequired

BinaryTree = TypedDict(
    "BinaryTree",
    {
        "value": str,
        "left": NotRequired["BinaryTree"],
        "right": NotRequired["BinaryTree"],
    },
)


def is_trees_synchronized(
    tree1: BinaryTree | None, tree2: BinaryTree | None
) -> list[bool | str]:
    result: list[bool | str] = [False, tree1["value"] if tree1 is not None else ""]

    if tree1 is not None and tree2 is not None:
        result[0] = tree1["value"] == tree2["value"]

        if "left" not in tree1 or "right" not in tree2:
            return result

        left_branch_01 = tree1["left"]
        right_branch_01: BinaryTree | None = tree1.get("right", None)

        left_branch_02: BinaryTree | None = tree2.get("left", None)
        right_branch_02 = tree2["right"]

        result[0] = (
            result[0]
            and is_trees_synchronized(tree1=left_branch_01, tree2=right_branch_02)[0]
        )

        result[0] = (
            result[0]
            and is_trees_synchronized(tree1=right_branch_01, tree2=left_branch_02)[0]
        )

    return result
